xcat-inventory/xcclient/inventory/dbfactory.py

```diff
@@ -46,77 +56,129 @@ def create_or_update(session,tabcls,key,newdict,ismatrixtable=True):
+           elif len(tabkeys)==1:
+               tabobjs=dbsession.query(tab).filter(getattr(tab,tabkeys[0]).in_(keys),or_(tab.disable == None, tab.disable.notin_(['1','yes']))).all()
+           elif len(tabkeys)>1:
+               for key in keys:
+                   if type(key)!=tuple:
+                       key=[key]
+                   kvdict=dict(zip(tabkeys,key))
+                   query=dbsession.query(tab)
+                   for key,value in kvdict.items():
+                       query=query.filter(getattr(tab,key).in_([value]))
+                   query=query.filter(or_(tab.disable == None, tab.disable.notin_(['1','yes'])))  
+                   tabobj=query.all()
+                   tabobjs.extend(tabobj)
+           if not tabobjs:
                continue
-           for myobj in tabobj:
+
+           objkeyname=tab.getobjkey()
+           dictoftab={}
+           for myobj in tabobjs:
                mydict=myobj.getdict()
-               mykey=mydict[tab.__tablename__+'.'+tabkey]
+
+               if len(objkeyname)==1:
+                   mykey=mydict[tab.__tablename__+'.'+objkeyname[0]]
+               elif len(objkeyname)>1:
+                   mykeylist=[]
+                   for key in objkeyname:
+                       mykeylist.append(mydict[tab.__tablename__+'.'+key])  
+                   mykey=tuple(mykeylist)
+
+               if mykey not in dictoftab.keys():
+                   dictoftab[mykey]=mydict
+               else:
+                   if type(dictoftab[mykey])!=list:
+                       dictoftab[mykey]=[dictoftab[mykey]]
+                   dictoftab[mykey].append(mydict)
+
+
+           for mykey in dictoftab.keys():
                if mykey not in ret.keys():
-                  ret[mykey]={}
-               ret[mykey].update(mydict)
-        return  ret 
+                   ret[mykey]={}
+               if type(dictoftab[mykey])==list:
+                   if tabname not in ret[mykey].keys():
+                       ret[mykey][tabname]=[]
+                   ret[mykey][tabname].extend(dictoftab[mykey])
+               else:
+                   ret[mykey].update(dictoftab[mykey])
+
+        for mykey in ret.keys():
+            if len(ret[mykey].keys())==1 and ret[mykey].keys()[0] in tabs :
+                ret[mykey]=ret[mykey][ret[mykey].keys()[0]]          
+
+        return ret 
 
     def settab(self,tabdict=None):
-        #print "=========matrixdbfactory:settab========"
+        #print("=========matrixdbfactory:settab========")
         #print(tabdict)
         #print("\n")
         if tabdict is None:
             return None
         for key in tabdict.keys():
             utils.verbose("  writting object: "+str(key),file=sys.stdout)
+            #clear any existing table entries before adding new entries
+            df=dbfactory(self._dbsession) 
+            df.cleartab(tabdict[key].keys(),[key])
```

Comment:
When it is not update, `manager.py` calls `cleartab` to clear up db, there is no data in DB now. In line 171 call cleartab, it clear up DB again in for loop. And if there are too many nodes, for each node, it  clear up `switch` table once, the performance can be enhanced here.

Reply:
will refine this in following sprints

Reply:
I opened a ticket to trace this enhance https://github.com/xcat2/xcat-inventory/issues/147, thanks